src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,146 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int NOT_FOUND = -1;
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            elements = (T[]) new Object[initialCapacity];
+        } else {
+            throw new IllegalArgumentException("Illegal initial capacity: " + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
+        growIfNeed();
 
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
 
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            growIfNeed();
+            T[] newElements = (T[]) new Object[elements.length];
+            newElements[index] = value;
+            System.arraycopy(elements, 0, newElements, 1, size);
+            elements = newElements;
+            size++;
+        } else {
+            growIfNeed();
+            T[] newElements = (T[]) new Object[elements.length];
+            System.arraycopy(elements, 0, newElements, 0, index);
+            newElements[index] = value;
+            System.arraycopy(elements, index, newElements, index + 1, size - index + 1);
+            elements = newElements;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            growIfNeed();
+            elements[size] = list.get(i);
+            size++;
+        }
```

Comment:
reuse existing methods
```suggestion
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```